### PR TITLE
Backport subdomain validation

### PR DESCRIPTION
**Pull request by sanmai**
This PR backports subdomain validation from #466, necessary to fix CVE-2021-30492.

I also added/fixed tests. Here's how you can run them: 
```
php vendor/bin/phpunit tests/Zendesk/API/UnitTests/InvalidSubdomainTest.php
php vendor/bin/phpunit tests/Zendesk/API/UnitTests/OrganizationsTest.php
```

All should be left is to tag a new release, update the Packagist registry, and tweak CVE-2021-30492 to mark this new version as unaffected.

### CODE REVIEW

File: tests/Zendesk/API/UnitTests/OrganizationsTest.php [hunk 1]
```diff
@@ -1,6 +1,6 @@
 <?php
```
Comment (ecoologic) [inline tests/Zendesk/API/UnitTests/OrganizationsTest.php:1]: This file does not exist in current master, and a similar issue has already been fixed, can you please try to merge master in here?

File: tests/Zendesk/API/UnitTests/InvalidSubdomainTest.php [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+ */
+class InvalidSubdomainTest extends BasicTest
+{
+    /**
+     * @expectedException InvalidArgumentException
+     */
+    public function testInvalidSubdomainThrows()
+    {
+        new Client('...', '');
+    }
```
Comment (ecoologic) [inline tests/Zendesk/API/UnitTests/InvalidSubdomainTest.php:19]: I'm afraid we would need a more specific test.

**Review by ecoologic — CHANGES_REQUESTED** [review]
Thank you very much for your contribution. ❤️ 

Please note the required changes.

File: tests/Zendesk/API/UnitTests/InvalidSubdomainTest.php (hunk 2 above)
Comment (sanmai) [reply tests/Zendesk/API/UnitTests/InvalidSubdomainTest.php:19]: I'm happy to add one. What test would you like to see?

**Comment by sanmai** [comment]
Alternatively, I'll be happy to close this PR as it looks like it is no longer necessary for my cause.

File: tests/Zendesk/API/UnitTests/GroupMembershipsTest.php [hunk 3]
```diff
@@ -1,6 +1,6 @@
 <?php
 
-namespace Zendesk\API\LiveTests;
+namespace Zendesk\API\UnitTests;
```
Comment (ecoologic) [inline tests/Zendesk/API/UnitTests/GroupMembershipsTest.php:3]: Thank you ❤️

File: tests/Zendesk/API/UnitTests/InvalidSubdomainTest.php (hunk 2 above)
Comment (ecoologic) [reply tests/Zendesk/API/UnitTests/InvalidSubdomainTest.php:19]: You're introducing a regex, could you please provide a test that passes the validation and one that throws `InvalidArgumentException`?

**Review by ecoologic — COMMENTED** [review]
Hi @sanmai  - thank you very much for your contribution and patience.

I know we've not given a reasonable feedback loop in the past, but moving forward, we are committed to maintaining this library and be nicer to our awesome open-source contributors.

Please keep in mind though, that I personally don't have a deep knowledge of this lib nor PHP, but I'll do my best in a world of conflicting priorities.

With that said, would you be so kind to address the comments?

Thanks a lot, long live open-source ❤️

File: tests/Zendesk/API/UnitTests/InvalidSubdomainTest.php (hunk 2 above)
Comment (sanmai) [reply tests/Zendesk/API/UnitTests/InvalidSubdomainTest.php:18]: Done!

File: tests/Zendesk/API/UnitTests/OrganizationsTest.php (hunk 1 above)
Comment (ecoologic) [reply tests/Zendesk/API/UnitTests/OrganizationsTest.php:1]: Thank you for the clarification.

**Review by ecoologic — APPROVED** [review]
Thank you for your contribution @sanmai.

We only maintain the latest version, but surely we welcome improving the value or older versions.

**Pull request merged**